From 5ccb4618a5e1f6beec429e45e599b9a998805130 Mon Sep 17 00:00:00 2001
From: Pete Beardmore <pete.beardmore@msn.com>
Date: Mon, 10 Mar 2014 14:51:10 +0000
Subject: [PATCH] alsa: Use card description in default sink/source prefix when
 available

-given an explicit device.description in card_properties, prefer this
information over other default prefixes (e.g. 'Built-in Audio') when
constructing sink/source descriptions
---
 src/modules/alsa/alsa-sink.c   |  3 +++
 src/modules/alsa/alsa-source.c |  3 +++
 src/pulsecore/card.h           |  1 +
 src/pulsecore/sink.c           | 10 +++++++---
 4 files changed, 14 insertions(+), 3 deletions(-)

diff --git a/src/modules/alsa/alsa-sink.c b/src/modules/alsa/alsa-sink.c
index 82c07c7..4cb9744 100644
--- a/src/modules/alsa/alsa-sink.c
+++ b/src/modules/alsa/alsa-sink.c
@@ -2255,6 +2255,9 @@ pa_sink *pa_alsa_sink_new(pa_module *m, pa_modargs *ma, const char*driver, pa_ca
             pa_proplist_sets(data.proplist, key, pa_proplist_gets(mapping->proplist, key));
     }
 
+    if (data.card)
+        if (pa_proplist_gets(data.card->proplist, PA_PROP_DEVICE_DESCRIPTION))
+            pa_proplist_sets(data.proplist, PA_PROP_DEVICE_CARD_DESCRIPTION, pa_proplist_gets(data.card->proplist, PA_PROP_DEVICE_DESCRIPTION));
     pa_alsa_init_description(data.proplist);
 
     if (u->control_device)
diff --git a/src/modules/alsa/alsa-source.c b/src/modules/alsa/alsa-source.c
index 58f00f7..3aac0dc 100644
--- a/src/modules/alsa/alsa-source.c
+++ b/src/modules/alsa/alsa-source.c
@@ -1975,6 +1975,9 @@ pa_source *pa_alsa_source_new(pa_module *m, pa_modargs *ma, const char*driver, p
             pa_proplist_sets(data.proplist, key, pa_proplist_gets(mapping->proplist, key));
     }
 
+    if (data.card)
+        if (pa_proplist_gets(data.card->proplist, PA_PROP_DEVICE_DESCRIPTION))
+            pa_proplist_sets(data.proplist, PA_PROP_DEVICE_CARD_DESCRIPTION, pa_proplist_gets(data.card->proplist, PA_PROP_DEVICE_DESCRIPTION));
     pa_alsa_init_description(data.proplist);
 
     if (u->control_device)
diff --git a/src/pulsecore/card.h b/src/pulsecore/card.h
index d72a293..7d9106b 100644
--- a/src/pulsecore/card.h
+++ b/src/pulsecore/card.h
@@ -47,6 +47,7 @@ typedef struct pa_card_profile {
 } pa_card_profile;
 
 #define PA_CARD_PROFILE_DATA(d) ((void*) ((uint8_t*) d + PA_ALIGN(sizeof(pa_card_profile))))
+#define PA_PROP_DEVICE_CARD_DESCRIPTION "device.card.description"
 
 struct pa_card {
     uint32_t index;
diff --git a/src/pulsecore/sink.c b/src/pulsecore/sink.c
index a8a91d6..2dc38e7 100644
--- a/src/pulsecore/sink.c
+++ b/src/pulsecore/sink.c
@@ -3416,9 +3416,13 @@ pa_bool_t pa_device_init_description(pa_proplist *p) {
     if (pa_proplist_contains(p, PA_PROP_DEVICE_DESCRIPTION))
         return TRUE;
 
-    if ((s = pa_proplist_gets(p, PA_PROP_DEVICE_FORM_FACTOR)))
-        if (pa_streq(s, "internal"))
-            d = _("Built-in Audio");
+    if (s = pa_proplist_gets(p, PA_PROP_DEVICE_CARD_DESCRIPTION))
+        d = s;
+
+    if (!d)
+        if ((s = pa_proplist_gets(p, PA_PROP_DEVICE_FORM_FACTOR)))
+            if (pa_streq(s, "internal"))
+                d = _("Built-in Audio");
 
     if (!d)
         if ((s = pa_proplist_gets(p, PA_PROP_DEVICE_CLASS)))
-- 
1.8.4

